### PR TITLE
feat(rides): adds rides resource group

### PR DESCRIPTION
**Pull request by codyleyhan**
# Description
- Adds full rides resource
- Added validation through go ozzo
- Addressed #8 

screen shot 2018-04-19 at 11 52 51 pm

# Steps To Reproduce
- Just run the server
- Hit the group of /api/v1/rides

# Tested
- Rides model unit tested
- Ride service unit tested
- Added a ride mock
screen shot 2018-04-19 at 11 54 14 pm

# Approved
- [x] Jahan
- [ ] Luke

### CODE REVIEW

**Comment by codyleyhan** [comment]
Already see that the date is not working correctly, will work on that

File: app.go [hunk 1]
```diff
@@ -94,8 +100,8 @@ func Start() {
 	pagesController.MountRoutes(app.Group(""))
 
 	auth := app.Group("/api/v1")
-
 	userController.MountRoutes(auth)
```
Comment (jcherianucla) [inline app.go]: Change to individual groups for each controller to decouple them.

File: models/ride.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	errs := ""
+
+	if r.DriverID == "" {
+		errs += "Driver ID: the ride needs a driver;"
+	}
+
+	if r.DriverID == "" {
+		errs += "Driver ID: the ride needs a driver;"
+	}
+
```
Comment (jcherianucla) [inline models/ride.go]: nit: Repeated if statement for DriverID

File: models/ride.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	if r.CarID == "" {
+		errs += "Car ID: the ride needs a car;"
+	}
+
+	if r.Seats < 0 {
+		errs += "seats: there needs to be a positive number of seats;"
+	}
```
Comment (jcherianucla) [inline models/ride.go]: nit: Car ID and Driver ID are capitalized but the others are not. Let's chose one or the other.

File: models/ride.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+		DriverID     string    `json:"driver_id" db:"driver_id"`
+		CarID        string    `json:"car_id" db:"car_id"`
+		Seats        int       `json:"seats" db:"seats"`
+		StartCity    string    `json:"start_city" db:"start_city"`
+		EndCity      string    `json:"end_city" db:"end_city"`
+		StartLat     float64   `json:"start_dest_lat" db:"start_dest_lat"`
+		StartLon     float64   `json:"start_dest_lon" db:"start_dest_lon"`
+		EndLat       float64   `json:"end_dest_lat" db:"end_dest_lat"`
+		EndLon       float64   `json:"end_dest_lon" db:"end_dest_lon"`
+		PricePerSeat int       `json:"price_per_seat" db:"price_per_seat"`
```
Comment (jcherianucla) [inline models/ride.go]: Prices could be floats? What if we want a price like $5.50 per seat.

File: models/ride.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+// ApplyUpdates attempts to update ride and validates the updates
+func (r *Ride) ApplyUpdates(o *RideChangeSet) error {
+	newRide := Ride{}
+
+	if err := mergo.Merge(&newRide, *r, mergo.WithOverride); err != nil {
+		return errors.New("There was a problem applying updates to ride")
+	}
+
```
Comment (jcherianucla) [inline models/ride.go]: No need for a Merge, just create a copy of the value of existing ride to newRide and update it accordingly. Removes extra dependency.

File: models/ride.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	err := newRide.Validate()
+	if err != nil {
+		return err
+	}
+
+	if err = mergo.Merge(r, newRide, mergo.WithOverride); err != nil {
+		return errors.New("There was a problem applying updates to ride " + err.Error())
+	}
+
```
Comment (jcherianucla) [inline models/ride.go]: Same as above with regards to mergo.

File: stores/postgres/ride.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+// Delete deletes the ride, does no verification
+func (r *RideStore) Delete(id string) error {
+	_, err := r.db.Exec(rideDeleteSQL, id)
+	return err
+}
+
+func (r *RideStore) getBy(query string, arg interface{}) (*models.Ride, error) {
+	var ride models.Ride
```
Comment (jcherianucla) [inline stores/postgres/ride.go]: Let's stick to ride := models.Ride{}

**Review by jcherianucla — CHANGES_REQUESTED** [review]
Most of it looks good but I want to try and revisit:
1.) Validation - Retry govalidator or another package
2.) Updates - I like the change set, but I don't like the crazy amount of if statements. Reflection could work, but tradeoffs in performance and readability to new devs. Maybe we should look into this a little more just to be confident.

**Review by jcherianucla — APPROVED** [review]

**Review by jcherianucla — APPROVED** [review]
Looks good to me

**Comment by epicchewy** [comment]
lgtm

**Comment by codyleyhan** [comment]
There were some changes so @epicchewy @jcherianucla  can you take a look again

**Review by jcherianucla — APPROVED** [review]
Looks lit to me

**Pull request merged**